### DISCUSSION TITLE
output folder options

### DISCUSSION THREAD

**Discussion by Katmoget** [Comfy-Org/ComfyUI#1292] [Q&A] [answered]
Hi,
Is there a way to change the default output folder ? I tried to add an output in the extra_model_paths.yaml file, the path gets added by ComfyUI on start up but it gets ignored when the png file is saved.
I'm not sure how to amend the folder_paths.py
I am using the WAS image save node in my own workflow but I can't always replace the default save image node with it in some complex workflow from other users. 
any advice ?
Thanks

**Comment by aegis72** [comment]
There's an add-on that is a more powerful save node. I am actually trying to figure out where it is from, since I found it by searching for "image" when clicking in the interface to add a node, and all the nodes are listed as top-level stuff. It looks like this: 
![image](https://github.com/comfyanonymous/ComfyUI/assets/118572301/9a3cc106-44f0-4a68-a413-d3ceadf5eb73)

It allows for a path change, as well as complex masking for the filename and other options.

And might be a simple replacement, even in complex workflows. Not sur eif that helps :( but maybe.

**Reply by Katmoget** [reply]
thanks, it looks like the WAS node that I mentioned, useful for small workflows but for complex ones it's not really practical to swap everywhere. 
Having the option to define the output path in a config file would be way better.

**Reply by Torcelllo** [reply]
Apart from the clever name and date node - I make a Symlink of my ComfyUI/Output folder so that the photos are stored on an external drive.

Delete or rename your ComfyUI Output folder (which for the sake of argument is C:\Comfyui\output).
Symlink format takes the "space" where this Output folder used to be and inserts a linked folder. The linked folder points to the new folder (say Drive X:\ComfyUI\output).

So using the CMD Window - mklink /D "C:\Comfyui\output" "X:\ComfyUI\output"

This will report that a succesful link has been made.
If you haven't deleted or renamed the original "C:\Comfyui\output" folder** - you will get an error!

This can take the burden off an overloaded C: Drive when hundreds and thousands of images pour out of ComfyUI each month!

**For ComfyUI_Windows_Portable - folder names are preceded with "ComfyUI_Windows_Portable\ComfyUI" so your output folder will be "ComfyUI_Windows_Portable\ComfyUI\output"

**Reply by NPDX909** [reply]
How do you get these options? My node has none of these.

**Reply by Jonseed** [reply]
> How do you get these options? My node has none of these.

Download the WAS Node Suite, and you'll find the Image Save node with all these options.

**Reply by ivica31** [reply]
is there one for video output?

**Reply by HHUHaoranGuo** [reply]
> Apart from the clever name and date node - I make a Symlink of my ComfyUI/Output folder so that the photos are stored on an external drive.除了聪明的名字和日期节点 - 我将 ComfyUI/Output 文件夹创建为符号链接，以便照片存储在外部驱动器上。
> 
> Delete or rename your ComfyUI Output folder (which for the sake of argument is C:\Comfyui\output).删除或重命名您的 ComfyUI 输出文件夹（假设为 C:\Comfyui\output）。 Symlink format takes the "space" where this Output folder used to be and inserts a linked folder. The linked folder points to the new folder (say Drive X:\ComfyUI\output).符号链接格式将占据这个输出文件夹曾经的位置，并插入一个链接文件夹。链接文件夹指向新的文件夹（例如驱动器 X:\ComfyUI\output）。
> 
> So using the CMD Window - mklink /D "C:\Comfyui\output" "X:\ComfyUI\output"使用命令提示符窗口 - mklink /D "C:\Comfyui\output" "X:\ComfyUI\output"
> 
> This will report that a succesful link has been made.已成功建立连接的汇报。 If you haven't deleted or renamed the original "C:\Comfyui\output" folder** - you will get an error!如果您没有删除或重命名原始的 "C:\Comfyui\output" 文件夹** - 您将收到错误信息！
> 
> This can take the burden off an overloaded C: Drive when hundreds and thousands of images pour out of ComfyUI each month!这可以在每月有成百上千张图片从 ComfyUI 流出时，减轻 C: 驱动器的负担！
> 
> **For ComfyUI_Windows_Portable - folder names are preceded with "ComfyUI_Windows_Portable\ComfyUI" so your output folder will be "ComfyUI_Windows_Portable\ComfyUI\output"

Very useful, thank you!

**Comment by Radtoo** [comment, marked as answer]
I also wished this was "the" comfyui application configuration file where paths can be configured, but as far as I know you can't. Can you use the startup parameter https://github.com/comfyanonymous/ComfyUI/issues/96#issuecomment-1498006080 ?

**Reply by Katmoget** [reply]
Thanks Radtoo!

As you suggested, following the advice in #96, ( I thought I had searched everywhere but clearly missed that one!) 
I managed to set the folder by adding the --output-directory in a batch file to startup comfy:

`python main.py  --output-directory D:\path_to_my_directory\`

**Reply by Kiaazad** [reply]
`python main.py  --output-directory D:\path_to_my_directory\`
Can I do the same for the temp and input directories?

**Reply by rafstahelin** [reply]
sorry, does this work? where did you edit code? in which file? thanks

**Reply by Katmoget** [reply]
Hi 

I created a batch file to start comfy, I'm using Windows, in the ComfyUI directory called Comfy_start.bat with this code  :

```
@echo off
git pull
python main.py --output-directory D:\SD\Save\ (replace with the path to your directory)


```
(you can comment out git pull if you don't want to run at every start)

That will change the default Comfy output directory to your directory every time you start comfy using this batch file. 
But if you want the files to be saved in a specific folder within that directory for example a folder automatically created per date you can do the following :

In your ComfyUI workflow
In the Save Image  node convert the filename_prefix to input, (right click on the text then select convert in the contextual menu)
then create a primitive node (add node>utils>primitive). Link the primitive to the filename_prefix.
and add your custom file name in the primitive field. 
I use %date:yyyy-MM-dd%/ComfyUI which creates a folder with the date in the directory I specified in my comfy_start.bat and save the file with the prefix ComfyUI

for more options see : https://www.andreszsogon.com/change-output-file-names-in-comfyui/ and https://blenderneko.github.io/ComfyUI-docs/Interface/SaveFileFormatting/

**Reply by ccarstens** [reply]
@Katmoget 
why are you running git pull after starting comfy? if you'd want to auto update on every start, it should go before if I'm not missing anything?

**Reply by Katmoget** [reply]
@ccarstens you're right should be :
@echo off
git pull
python main.py --output-directory D:\SD\Save\ (replace with the path to your directory)

**Reply by aegis72** [reply]
When I attempt to do the same for my models directory, I get a permission denied error. Anyone else encounter that and know the fix? I'm trying to externalize my models directory rather than share with A1111, which I no longer use. I am using ComfyUI portable on windows. (in my case python main.py --extra-model-paths-config h:\models)

**Reply by Katmoget** [reply]
@aegis72 you need to edit an rename the extra_model_paths.yaml.example in the ComfyUI directory to change the path to your models folder.

**Comment by aegis72** [comment]
Yes… the issue is when I do that, I get the permission error. I’ve tried
both ways, command line switch and yaml, but get that same message on each
method.

**Reply by Katmoget** [reply]
seems to be a windows permission problem, if you haven't already have a look at https://github.com/comfyanonymous/ComfyUI/issues/1585#issuecomment-1732240398
and there is also a open issue : https://github.com/comfyanonymous/ComfyUI/issues/851

**Comment by serget2** [comment]
I wish there was a way to add used model and date to the name automatically, so when I am trying the same prompt but all the different models it just autoadds that to the name

**Comment by TwoPintOhh** [comment]
Look in folder_paths.py and you can change it there, just comment out the original

#output_directory = os.path.join(os.path.dirname(os.path.realpath(__file__)), "output")
#temp_directory = os.path.join(os.path.dirname(os.path.realpath(__file__)), "temp")

output_directory = "R:\Output"
temp_directory = "R:\Temp"

**Reply by screan** [reply]
> Look in folder_paths.py and you can change it there, just comment out the original
> 
> #output_directory = os.path.join(os.path.dirname(os.path.realpath(**file**)), "output") #temp_directory = os.path.join(os.path.dirname(os.path.realpath(**file**)), "temp")
> 
> output_directory = "R:\Output" temp_directory = "R:\Temp"

tried this, still same  error:

**** ERROR: Saving image outside the output folder is not allowed.

**Reply by ltdrdata** [reply]
Don't modify code. There are `--output-directory` and `--temp-directory` cli options.
Code modification will blocking updating of ComfyUI.

**Comment by dekdes** [comment]
yes i can confirm that this worked: 

i just edited the run_nvidia_gpu.bat  file  

original was
.\python_embeded\python.exe -s ComfyUI\main.py --windows-standalone-build 
pause 

changed LINE 1 to 

.\python_embeded\python.exe -s ComfyUI\main.py --windows-standalone-build --output-directory E:\01AI\Comfy_output

**Reply by OrinZ** [reply]
This solution only works for paths that do not include spaces. Having spaces breaks the restart function, even with quotes. Unexpected behavior but it's what I've observed.

**Reply by Jonseed** [reply]
I have also noted that paths with spaces breaks.
